ds701_book/27-RNN.qmd

```diff
@@ -0,0 +1,640 @@
+- **Less Flexibility**: GRUs offer less flexibility in terms of controlling the flow of information through the network.
+
+# RNN Examples
+
+## Appliance Energy Prediction
+
+In this example we will use a [dataset](https://archive.ics.uci.edu/dataset/374/appliances+energy+prediction) containing the energy usage in Watt hours (Wh) of appliances in a low energy building.
+
+In addition to the appliance energy information, the dataset includes the house temperature and humidity conditions. We will build an LSTM model that predicts the energy usage of the appliances.
+
+## Load the Data
+
+The code cells below load the dataset.
+
+```{python}
+#| code-fold: false
+import os
+
+file_path = 'energydata_complete.csv'
+url = "https://archive.ics.uci.edu/ml/machine-learning-databases/00374/energydata_complete.csv"
+
+if os.path.exists(file_path):
+    data = pd.read_csv(file_path)
+else:
+    data = pd.read_csv(url)
+    data.to_csv(file_path, index=False)
+
+data.head()
+```
+
+## Resample Data
+
+We're interested in the `Appliances` column, which is the energy use of the appliances in Wh. 
+
+First, we'll resample the data to hourly resolution and fill missing values using the forward fill method.
+
+```{python}
+data['date'] = pd.to_datetime(data['date'])
+data.set_index('date', inplace=True)
+
+data = data['Appliances'].resample('h').mean().ffill() # Resample and fill missing
+
+data.head()
+```
+
+## Preparing the Data
+
+We create train-test splits and scale the data accordingly. 
+
+In addition, we create our own dataset class. In this class we create lagged sequences of the energy usage data. The amount of lag, or look-back sequence length, is set to 24.
+
+```{python}
+train_size = int(len(data) * 0.8)
+test_size = len(data) - train_size
+
+train_data = data[:train_size]
+test_data = data[train_size:]
+
+# Normalize data
+scaler = MinMaxScaler()
+train_data_scaled = scaler.fit_transform(train_data.values.reshape(-1, 1))
+test_data_scaled = scaler.transform(test_data.values.reshape(-1, 1))
+
+# Prepare data for LSTM
+class TimeSeriesDataset(Dataset):
+    def __init__(self, data, seq_length):
+        self.data = data
+        self.seq_length = seq_length
+
+    def __len__(self):
+        return len(self.data) - self.seq_length
+
+    def __getitem__(self, index):
+        X = self.data[index:index + self.seq_length]
+        y = self.data[index + self.seq_length]
+        return torch.tensor(X, dtype=torch.float32), torch.tensor(y, dtype=torch.float32)
+
+seq_length = 24
+train_dataset = TimeSeriesDataset(train_data_scaled, seq_length)
+test_dataset = TimeSeriesDataset(test_data_scaled, seq_length)
+
+# Create DataLoader
+train_loader = DataLoader(train_dataset, batch_size=8, shuffle=True)
+test_loader = DataLoader(test_dataset, batch_size=8, shuffle=False)
+```
+
+## LSTM Architecture
+
+We define our LSTM architecture in the code cell below.
```

Comment:
I wonder if there is a graph visualization tool that can show a few of the repeating cells...

Reply:
I attempted to draw this using both torchviz and torchview and it looks like this. I'm not sure it's worth showing.
![LSTMModel](https://github.com/user-attachments/assets/4ee622f7-a0e3-4f94-b713-990e70837f6c)

Reply:
yeah, that is too detailed.